docs/apis/streaming/windows.md

```diff
@@ -24,1023 +24,608 @@ specific language governing permissions and limitations
-        The elements on the triggered window are henceforth discarded.
-      </p>
-{% highlight scala %}
-windowedStream.trigger(ProcessingTimeTrigger.create);
-{% endhighlight %}
-    </td>
-  </tr>
-  <tr>
-    <td><strong>Watermark trigger</strong></td>
-    <td>
-      <p>
-        A window is fired when a watermark with value that exceeds the window's end-value has been received.
-        The elements on the triggered window are henceforth discarded.
-      </p>
-{% highlight scala %}
-windowedStream.trigger(EventTimeTrigger.create);
-{% endhighlight %}
-    </td>
-  </tr>
-  <tr>
-    <td><strong>Continuous processing time trigger</strong></td>
-    <td>
-      <p>
-        A window is periodically considered for being fired (every 5 seconds in the example).
-        The window is actually fired only when the current processing time exceeds its end-value.
-        The elements on the triggered window are retained.
-      </p>
-{% highlight scala %}
-windowedStream.trigger(ContinuousProcessingTimeTrigger.of(Time.seconds(5)));
-{% endhighlight %}
-    </td>
-  </tr>
-  <tr>
-    <td><strong>Continuous watermark time trigger</strong></td>
-    <td>
-      <p>
-        A window is periodically considered for being fired (every 5 seconds in the example).
-        A window is actually fired when a watermark with value that exceeds the window's end-value has been received.
-        The elements on the triggered window are retained.
-      </p>
-{% highlight scala %}
-windowedStream.trigger(ContinuousEventTimeTrigger.of(Time.seconds(5)));
-{% endhighlight %}
-    </td>
-  </tr>
-  <tr>
-    <td><strong>Count trigger</strong></td>
-    <td>
-      <p>
-        A window is fired when it has more than a certain number of elements (1000 below).
-        The elements of the triggered window are retained.
-      </p>
 {% highlight scala %}
-windowedStream.trigger(CountTrigger.of(1000));
+val input: DataStream[(String, Long)] = ...
+
+input
+    .keyBy(<key selector>)
+    .window(<window assigner>)
+    .apply(new MyWindowFunction())
+
+/* ... */
+
+class MyWindowFunction extends WindowFunction[(String, Long), String, String, TimeWindow] {
+
+  def apply(key: String, window: TimeWindow, input: Iterable[(String, Long)], out: Collector[String]): () = {
+    var count = 0L
+    for (in <- input) {
+      count = count + 1
+    }
+    out.collect(s"Window $window count: $count")
+  }
+}
 {% endhighlight %}
-    </td>
-  </tr>
-  <tr>
-    <td><strong>Purging trigger</strong></td>
-    <td>
-      <p>
-        Takes any trigger as an argument and forces the triggered window elements to be
-        "purged" (discarded) after triggering.
-      </p>
-{% highlight scala %}
-windowedStream.trigger(PurgingTrigger.of(CountTrigger.of(1000)));
+</div>
+</div>
+
+### WindowFunction with Incremental Aggregation
+
+A `WindowFunction` can be combined with either a `ReduceFunction` or a `FoldFunction`. This allows
```

Comment:
How about this:

A `WindowFunction` can be combined with either a `ReduceFunction` or a `FoldFunction`. When doing
this, the `ReduceFunction`/`FoldFunction` will be used to incrementally aggregate elements as they
arrive while the `WindowFunction` will be provided with the aggregated result when the window is
ready for processing. This allows to get the benefit of incremental window computation and also have the additional meta information that writing a `WindowFunction` provides.